### DISCUSSION TITLE
component markers

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#1375] [Ideas & RFCs]
Imagine we allow components to be attached to marker characters, eg

```ftd
-- import: foo

-- ftd.text: hello 💃world💃
$marker:💃$: foo.bold
```

Which is equivalent to:

```ftd
-- import: foo

-- ftd.text: hello {foo.bold: world}
```

**Comment by siddhantk232** [comment]
I think that the syntax of this can be improved by https://github.com/orgs/fastn-stack/discussions/884

**Reply by amitu** [reply]
@siddhantk232 I think you linked to wrong discussion here.

**Reply by siddhantk232** [reply]
oh no I copied the wrong link. Lemme see if I can find that discussion

**Reply by siddhantk232** [reply]
fixed the link

**Comment by siddhantk232** [comment]
We could use TOML's dotted keys here to get the following syntax:

```ftd
-- import: foo

-- ftd.text: hello 💃world💃
marker: $foo.bold
marker.key: 💃
```

Equivalent to:

```
-- import: foo

-- ftd.text: hello {-- foo.bold: world}
```

**Reply by amitu** [reply]
How do you add more than one marker?

**Reply by siddhantk232** [reply]
I am not sure. Let's suppose `marker` is a record:

```ftd
-- record marker:
string key:
component with:
```

and `ftd.text` is able to take `marker` somehow:

```
-- ftd.text: some text 💃bold💃 😃emphasised😃
markers: { ;; notice the plural `markers`. It's a list
  .key: 💃
  .with: foo.bold ;; f-script so $ is omitted in $foo.bold
}
markers: {
  .key: 😃
  .with: foo.em
} ;; these two header values will be UNIONED together to create a `marker list markers`
```

We created a record instance (of record `marker`) on the fly and passed it to header `markers` of `ftd.text` component. The other version where record instance was created separately would have been:

```ftd

-- marker list my-markers:

-- marker:
key: <>
with: <>

-- marker:
key: <>
with: <>

-- end: my-markers

-- ftd.text: some text 💃bold💃 😃emphasised😃
markers: $my-markers
```


Can we say that: "if the header is a list type then multiple declarations of that header in component call will be a union operation on the value of said headers?"

definition of `ftd.text`:
```ftd
-- component text:
marker list markers:

-- end: text
```

**Reply by siddhantk232** [reply]
I don't like this. This has to be a compile time transformation (or parse time transformation) so `marker` header can't really exist on the `ftd.text` component.

**Reply by amitu** [reply]
> Can we say that: "if the header is a list type then multiple declarations of that header in component call will be a union operation on the value of said headers?"

We did that in ftd 0.2 I think. Not sure if we removed it for reasons, or if it got lost during 0.3 somehow. 

Do you think it is a good idea? It does save one reason to use `{}`.

**Reply by siddhantk232** [reply]
Let's see:

This is when union of both header declarations will happen

```ftd
-- ftd.text: some text 💃bold💃 😃emphasised😃
markers: {
  .key: 💃
  .with: foo.bold
}
markers: {
  .key: 😃
  .with: foo.em
}
```

no union but we have ability to create variables on the fly (a list of `marker` in this case).
```ftd
-- ftd.text: some text 💃bold💃 😃emphasised😃
markers: { ;; notice the plural `markers`. It's a list of `marker`
  -- marker:
  key: 💃
  with: foo.bold ;; f-script so $ is omitted in $foo.bold
 
 -- marker:
  key: <>
  with: <>
}
```

It's not a big deal to allow union of same headers but that is a behaviour that we will have to communicate to the users. I feel like the second code snippet is better as it's more explicit.

**Reply by amitu** [reply]
In the first snippet, I am not sure the .key: syntax is necessary. We could have just used key:

**Comment by amitu** [comment]
I am also happy to give up on inline styling related concerns. 

<img width="855" alt="image" src="https://github.com/user-attachments/assets/84f508a0-708f-4cf4-aa8e-f3e4af332a92">

<img width="657" alt="image" src="https://github.com/user-attachments/assets/5818f581-edfd-47a6-be97-dd1f53fccc88">


These are the motivation for them. But we have plenty of inline styling available in markdown, strong, emphasis, strikethrough, highlight, super script, subscript. What if we "appropriate" those, and allow people to say highlight in this text node will be done by changing the color and strong means circle the word etc.
